README.md
# harmoniask

Exploring jazz music theory with Haskell.

Harmoniask is a little project to embed elements of Jazz music into Haskell, such as chords, substitutions, standard harmonic sequences and so on. 

As a standalone application, it can read a text file containing sequences of chords and output it in a MIDI file, possibly using alternative tunings (eg. Pythagoric, Werckmeister...). Note however that not every MIDI player support the MTS (Midi Tuning Standard) for alternative tunings.

### How to install

	git clone
	stack build
	
### Some examples

A simple chord, both in its abstract and concrete representations: 

	λ> majTriad
	Chord [Unison,ThirdMaj,Fifth]
	
	λ> realizeChord majTriad 60 -- 60 = Middle C
	[60,64,67]

Constructing chords via combinators:

	λ> addOnTop SeventhMin majTriad 
	Chord [Unison,ThirdMaj,Fifth,SeventhMin]
	
Constructing more complex chords:

	λ> x_9b = 
		addOnTop SeventhMin |> 
		addOnTop NinthMaj   |> 
		alter NinthMaj Flat |> 
		remove Fifth
	
	λ> x_9b majTriad 
	Chord [Unison,ThirdMaj,SeventhMin,NinthMin]
	
	λ> realizeChord (x_9b majTriad) 60 
	[60,64,70,72]

Some simple `Sequence` (sequence of chords that stays inside a tonality):

	λ> v_i = Sequence [Fifth, Unison] [x_7 majTriad, x_maj7 majTriad]
	Sequence [Fifth,Unison] 
	         [Chord [Unison,ThirdMaj,Fifth,SeventhMin]
			 ,Chord [Unison,ThirdMaj,Fifth,SeventhMaj]]	
	
	λ> realizeSeq 60 v_i
	[[67,71,74,77],[60,64,67,71]] -- G7 | Cmaj7

Some `Sequence` combinators:

	λ> ii_v_i = Sequence 
	            [SecondMaj, Fifth, Unison]
                [x_7 minTriad, x_7 majTriad, x_maj7 majTriad]
	
	λ> ii_v_tritone = (removeLastSeq |>>| tritoneSub)
	λ> ii_v_tritone ii_v_i
	Sequence [SecondMaj,SecondMin] 
		     [Chord [Unison,ThirdMin,Fifth,SeventhMin], 
			  Chord [Unison,ThirdMaj,Fifth,SeventhMin]]
			  
	λ> realizeSeq 60 $ ii_v_tritone ii_v_i
	[[62,65,69,72],[61,65,68,71]] -- D-7 | Db7

Realizing a `Sequence` while trying to reduce the distance between the notes of subsequent chords: 

	λ> realizeSeqCompact 60 ii_v_i
	[[62,65,69,72], -- D-7
	 [62,65,67,71], -- G7, 2nd inversion
	 [60,64,67,71]] -- Cmaj7
	
Creating `Harmony`s, a collection of sequences in different tonalities:

	circleFifths :: Harmony
	circleFifths = Harmony intervals sequences
		where intervals = concat $ replicate 6 [Up Fourth, Down Fifth]
		sequences       = replicate 12 ii_v_i

	> circleFifths
	Harmony 
		[Up Fourth, Down Fifth, Up Fourth, Down Fifth...] 
		[Sequence [SecondMaj,Fifth,Unison] 
		          [Chord [Unison,ThirdMin,Fifth,SeventhMin],
				   Chord [Unison,ThirdMaj,Fifth,SeventhMin],
				   Chord [Unison,ThirdMaj,Fifth,SeventhMaj]],
	     Sequence [SecondMaj,Fifth,Unison] 
		          [Chord [Unison,ThirdMin,Fifth,SeventhMin],
				   Chord [Unison,ThirdMaj,Fifth,SeventhMin],
				   Chord [Unison,ThirdMaj,Fifth,SeventhMaj]],
	     Sequence ... ... ... ]
		 
Exporting to a Midi file:

	λ> examplePitches = realizeHarmonyWith (noteToMidi C) realizeSeqCompact circleFifths
	λ> exampleTrack = makeTrack (MidiOptions 120 Nothing Nothing) examplePitches
	λ> writeMidiFile "examples/out.mid" $ createMidiFile $ exampleTrack

[Here](examples/out.mid) is the result.




ChangeLog.md
# Changelog for harmoniask

## Unreleased changes


test/Main.hs
import Test.Hspec
import Test.QuickCheck
import Control.Exception (evaluate)
import Control.Monad (zipWithM_)
import Data.Bits as DB
import Harmoniask.Tunings as HT

main :: IO ()
main = hspec $ do
  describe "HT (Harmoniask Tuning)" $ do
    zipWithM_ temperamentTest [eqHz, pytHz, werck1Hz, strangeHz]
                              ["Equable", "Pythagoric", "Werckmeister", "Strange"]

  describe "HT.midiFineTune, HT.freqsToCents, Ht.transform" $ do
    it "<distance in cents> -> <mst format> -> <distance in cents> = ID" $
      prop_cents

------------------------------------------------------

-- Test some properties of frequencies-pitch map
temperamentTest :: Temperament -> String -> SpecWith ()
temperamentTest temp name = describe name $ do
  it "has all the keys of a piano, from midi 21 to 104 (A0 to G#7)" $ do
    map fst pitchFreqs `shouldBe` [21..104]

  it "has all Word8 new base pitch values under 127" $ do
    mstMap `shouldSatisfy` (null . filter (\note -> (note !! 1) > 127))

  it "has all Word8 MSB cents values under 127" $ do
    mstMap `shouldSatisfy` (null . filter (\note -> (note !! 2) > 127))

  it "has all Word8 LSB cents values under 127" $ do
    mstMap `shouldSatisfy` (null . filter (\note -> (note !! 3) > 127))
    
  where pitchFreqs = HT.pianoRange temp
        mstMap     = HT.makeMap pitchFreqs


-- Test that, given two frequencies f1 and f2, going from their distance
-- in cents to their distance expressed in MST format to (again) their
-- cents distance gives the starting distance (modulo some floating
-- point approximations).
between :: Ord a => a -> a -> a -> Bool
between x a b = x > a && x < b

prop_cents :: Property
prop_cents = 
  forAll (suchThat arbitrary (\x -> x > 27.5)) $ \f1 ->
  forAll (suchThat arbitrary (\x -> (x > 27.5) &&
                               between (freqsToCents f1 x) 0 100)) $ \f2 ->
  (prop_cents_internal f1 f2)                                                                                          


prop_cents_internal :: Double -> Double -> Property
prop_cents_internal f1 f2 =  property $  
  (fromIntegral bitwise * 0.0061) `within` centsDist 
  where centsDist  = freqsToCents f1 f2        
        [msb, lsb] = map (\x -> (fromIntegral x) :: Int) $ midiFineTune centsDist

        bitwise :: Int
        bitwise = fromIntegral $ ((DB.shift msb 7) DB..|. (lsb DB..&. 127))

        within :: Real a => a -> a -> Bool
        within a b = abs (a - b) <= 1


test/Tests.hs
import Test.Hspec
import Test.QuickCheck
import Control.Monad (zipWithM_)
import Data.Maybe (isNothing, fromJust)
import ChordParserTest as CPT
import TuningTest as TT

main :: IO ()
main = sequence_ [CPT.mainChordParserTest, TT.mainTuningTest]
  


test/TuningTest.hs
module TuningTest where

import Test.Hspec
import Test.QuickCheck
import Control.Exception (evaluate)
import Control.Monad (zipWithM_)
import Data.Bits as DB
import Harmoniask.Tunings as HT


mainTuningTest :: IO ()
mainTuningTest = hspec $ do
  describe "HT (Harmoniask Tuning)" $ do
    zipWithM_ temperamentTest [eqHz, pytHz, werck1Hz, strangeHz]
                              ["Equable", "Pythagoric", "Werckmeister", "Strange"]

  describe "HT.midiFineTune, HT.freqsToCents, Ht.transform" $ do
    it "<distance in cents> -> <mst format> -> <distance in cents> = ID" $
      prop_cents

------------------------------------------------------

-- Test some properties of frequencies-pitch map
temperamentTest :: Temperament -> String -> SpecWith ()
temperamentTest temp name = describe name $ do
  it "has all the keys of a piano, from midi 21 to 104 (A0 to G#7)" $ do
    map fst pitchFreqs `shouldBe` [21..104]

  it "has all Word8 new base pitch values under 127" $ do
    mstMap `shouldSatisfy` (null . filter (\note -> (note !! 1) > 127))

  it "has all Word8 MSB cents values under 127" $ do
    mstMap `shouldSatisfy` (null . filter (\note -> (note !! 2) > 127))

  it "has all Word8 LSB cents values under 127" $ do
    mstMap `shouldSatisfy` (null . filter (\note -> (note !! 3) > 127))
    
  where pitchFreqs = HT.pianoRange temp
        mstMap     = HT.makeMap pitchFreqs


-- Test that, given two frequencies f1 and f2, going from their distance
-- in cents to their distance expressed in MST format to (again) their
-- cents distance gives the starting distance (modulo some floating
-- point approximations).
between :: Ord a => a -> a -> a -> Bool
between x a b = x > a && x < b

prop_cents :: Property
prop_cents = 
  forAll (suchThat arbitrary (\x -> x > 27.5)) $ \f1 ->
  forAll (suchThat arbitrary (\x -> (x > 27.5) &&
                               between (freqsToCents f1 x) 0 100)) $ \f2 ->
  (prop_cents_internal f1 f2)                                                                                          


prop_cents_internal :: Double -> Double -> Property
prop_cents_internal f1 f2 =  property $  
  (fromIntegral bitwise * 0.0061) `within` centsDist 
  where centsDist  = freqsToCents f1 f2        
        [msb, lsb] = map (\x -> (fromIntegral x) :: Int) $ midiFineTune centsDist

        bitwise :: Int
        bitwise = fromIntegral $ ((DB.shift msb 7) DB..|. (lsb DB..&. 127))

        within :: Real a => a -> a -> Bool
        within a b = abs (a - b) <= 1


app/Main.hs
module Main where

import Harmoniask
import Harmoniask.Parser as HP
import Harmoniask.Tunings as HT
import Data.Word (Word32)
import Data.List (intercalate)
import Options.Applicative

data MyApp = MyApp { inputFile  :: String
                   , outputFile :: String
                   , bpm        :: Word32
                   , temperamentName :: HT.DefaultTuning
                   , shortestPath :: Bool }

main :: IO ()
main = execParser options >>= runHarmoniaskWithOptions
  where
    parser = MyApp
      <$> argument str (metavar "INPUT"
                       <> help "Input file; a textual document where chords are separated by spaces")

      <*> argument str (metavar "OUTPUT"
                       <> help "Output file; a midi file.")

      <*> argument auto (metavar "BPM"
                       <> value 80
                       <> help "Tempo of the output midi file." )

      <*> option auto (metavar "Temperament"
                       <> short 't'
                       <> long "temperament"
                       <> value HT.Equable
                       <> help
                          (intercalate ", " $ map show $ ([minBound..maxBound] :: [HT.DefaultTuning])))

      <*> switch (short 's'
                       <> long "shortest"
                       <> help "Not implemented yet.")
      
    options = info (parser <**> helper)
      (fullDesc <>
       progDesc "Given a textual file INPUT containing chords, returns an harmonized midi file OUTPUT" <>
       header "harmoniask - chords harmonizer")


----------------------------

runHarmoniaskWithOptions :: MyApp -> IO ()
runHarmoniaskWithOptions options =
  application (inputFile options)
              (outputFile options)
              (MidiOptions (Main.bpm options)
                            Nothing
                            temperament')
  where temperament' = HT.temperament $ temperamentName options
  
application :: String -> String -> MidiOptions -> IO ()
application fnameIn fnameOut options = do
  rawInput <- readFile fnameIn
  
  let chords = map HP.stringToPitches $ words rawInput
  
  writeMidiFile fnameOut $
    createMidiFile $
      makeTrack options chords



src/HarmoniaskParser.hs
module HarmoniaskParser where

import Harmoniask
import Text.ParserCombinators.ReadP
import Data.Maybe (fromJust)

readAccident :: Char -> Maybe Accident
readAccident 'b' = Just Flat
readAccident '#' = Just Sharp
readAccident '♭' = Just Flat
readAccident _   = Nothing

data Tonality = Major | Minor
  deriving (Eq, Show)

data ColorNote = ColorNote Interval (Maybe Accident)
  deriving (Eq, Show)
readColorNoteDiatonic :: String -> Interval
readColorNoteDiatonic "4" = Interval Fourth
readColorNoteDiatonic "5" = Interval Fifth
readColorNoteDiatonic "9" = Interval NinthMaj
readColorNoteDiatonic "11" = Interval Eleventh
readColorNoteDiatonic "13" = Interval ThirteenthMaj
readColorNoteDiatonic x    = error "This should not happen"

data ColorForm = Sus2 | Sus4 | Simple7 | Maj7
  deriving (Eq, Show)
readColorForm :: String -> ColorForm
readColorForm "sus2" = Sus2
readColorForm "sus4" = Sus4
readColorForm "7" = Simple7
readColorForm "maj7" = Maj7


data ParsedChord = ParsedChord
  { tonik      :: (Note, Maybe Accident)
  , tonality   :: Tonality
  , colorForm  :: ColorForm
  , colorNotes :: [ColorNote] 
  } deriving (Eq, Show)


parsedToPitches :: ParsedChord -> [Pitch]
parsedToPitches (ParsedChord (tonic, mayAcc) tonality form colorNotes) =
  realizeChord chordIntervals tonicPitch
  
  where tonicPitch = noteToMidi tonic +
          (case mayAcc of
              Nothing -> 0
              Just Sharp -> 1
              Just Flat  -> -1)
        baseChord = case tonality of
          Major -> majTriad
          Minor -> minTriad
        specialForm = case form of
          Sus2 -> sus4
          Sus4 -> sus4
          Maj7 -> x_maj7
          Simple7 -> x_7
        colNs = map (\(ColorNote i acc) ->
                       if acc == Nothing
                       then i
                       else alter i (fromJust acc)) colorNotes
        addNoteComb = foldr (|>) id (map addOnTop colNs)
        chordIntervals = (addNoteComb |> specialForm) baseChord
          
        
------------------------------------------

isDiatonicNote :: Char -> Bool
isDiatonicNote char =
  any (char ==) "ABCDEFGabcdefg"

isAccident :: Char -> Bool
isAccident char =
  any (char ==) "b#♭"


-- | C#-maj7b9b13
--   ^^ 
tonicParser :: ReadP (Note, Maybe Accident)
tonicParser = do
  n <- fmap (\x -> (read [x]) :: Note) $ satisfy isDiatonicNote
  a <- option ' ' (satisfy isAccident)

  return (n, readAccident a)

-- | C#-maj7b9b13
--     ^ 
tonalityParser :: ReadP Tonality
tonalityParser = do
  ton <- option '+' (satisfy (\c -> any (c ==) "+-"))
  return $ if ton == '+' then Major else Minor

-- | C#-maj7b9b13
--          ^^^^^
colorNoteParser :: ReadP ColorNote
colorNoteParser = do
  a <- option ' ' (satisfy isAccident)
  n <- fmap readColorNoteDiatonic $ choice $ map string ["4", "5", "9", "11", "13"]
  return $ ColorNote n (readAccident a)

-- | C#-maj7b9b13
--      ^^^^
colorFormParser :: ReadP ColorForm
colorFormParser =
  fmap readColorForm $ choice $ map string $ ["7", "maj7", "sus2", "sus4"]

  

chord :: ReadP ParsedChord
chord = do
  tonik <- tonicParser
  ton   <- tonalityParser
  form  <- colorFormParser
  cols  <- many colorNoteParser
  return $ ParsedChord tonik ton form cols
  
stringToPitches :: String -> [Pitch]
stringToPitches = parsedToPitches . fst . last . readP_to_S chord

stringToPitches2 :: String -> Maybe [Pitch]
stringToPitches2 s = fmap parsedToPitches $ result $ readP_to_S chord s
  where result [] = Nothing
        result xs = Just $ fst $ last xs

stringToPitches22 :: String -> Maybe [Pitch]
stringToPitches22 s = fmap parsedToPitches $ result $ readP_to_S chord s
  where result [] = Nothing
        result xs = Just $ fst $ last xs
        
stringToPitches3 :: String -> Either String [Pitch]
stringToPitches3 s = fmap parsedToPitches $ result $ readP_to_S chord s
  where result [] = Left $ "Parser could not parse the chord " ++ s
        result xs = Right $ fst $ last xs


src/Harmoniask.hs
{-# LANGUAGE NumericUnderscores #-}

module Harmoniask where

import ZMidi.Core
import Data.List (sort, elemIndex, concatMap)
import Data.Maybe (fromJust, isNothing)
import GHC.Word
import Harmoniask.Tunings as HT

{- |
A typeclass for things that can be jumped from.

It is useful because of the distinction between a "static" interval (`Interval`) and
a "dynamic" interval (`MelodicInterval`, that can represent an interval going up or down).

@jump@ jumps from a MIDI pitch by a distance of an interval.
-}
class MusicalInterval a where
  jump :: Pitch -> a -> Pitch

  
{- |
This datatype codifies interval distances.

It can be thought as a "static" interval, that is one that doesn't encode movement information.
It is used for `Chord`s and `Sequence`s tonics (see later).
-}
data Interval = Unison | SecondMin | SecondMaj | ThirdMin | ThirdMaj | Fourth | Tritone | Fifth | SixthMin | SixthMaj | SeventhMin | SeventhMaj | NinthMin | NinthMaj | TenthMin | TenthMaj | Eleventh | OctaveTritone | Twelfth | ThirteenthMin | ThirteenthMaj | DoubleSeventhMin | DoubleSeventhMaj | DoubleOctave
  deriving (Show, Eq, Enum, Ord)

{- |
By default, a jump of an `Interval` is a jump up.
-}
instance MusicalInterval Interval where
  jump note interval = note + fromEnum interval

  
{- |
This datatype codifies melodic intervals.

It can be thought as a "dynamic" interval, that is one that encodes movement informations
(either upward `Up` or downward `Down`).

It is used for constructing melodies "out of context" (aka. melodic lines not bounded to
a tonal center such as a tonic).
-}
data MelodicInterval =
    Up Interval   -- ^ An interval jumping up.
  | Down Interval -- ^ An interval jumping down.
  deriving (Show, Eq)

instance Enum MelodicInterval where
  fromEnum (Up i) = fromEnum i
  fromEnum (Down i) = - (fromEnum i)

  toEnum n
    | n >= 0    = Up $ toEnum n
    | otherwise = Down $ toEnum (-n)
  
instance Ord MelodicInterval where
  compare (Up x) (Up y)     = compare x y
  compare (Down x) (Down y) = compare y x

  compare (Up _) _   = GT
  compare (Down _) _ = LT

instance MusicalInterval MelodicInterval where
  jump note (Up interval)   = note + fromEnum interval
  jump note (Down interval) = note - fromEnum interval
  

{- |
Used only for convenience, not really important.
-}
data Note = C | Cs | Db | D | Ds | Eb | E | F | Fs | Gb | G | Gs | Ab | A | As | Bb | B
  deriving (Show, Eq, Read)

{- |
MIDI Pitch, in range 0-128.
-}
type Pitch = Int -- Midi pitch (C4 = 60) (da cambiare in Word8?)


data Accident = Sharp | Flat deriving (Show, Eq)



-- ** Utility functions

mapEq :: (a -> Bool) -> (a -> a) -> [a] -> [a]
mapEq is f = map (\x -> if is x then f x else x)

-- | Rotates a list and applies a function to the piece of list pasted at the end.
rotateWith :: Ord a => (a -> a) -> Int -> [a] -> [a]
rotateWith func n xs = sort $ newStart ++ (map func newEnd) 
  where newEnd   = take n xs
        newStart = drop n xs


-- * Notes and intervals

-------------------------
-- Notes and intervals --
-------------------------
midiToNote x = names !! ((x - 24) `rem` 12)
  where names = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"]

noteToMidi :: Note -> Pitch
noteToMidi C = 60
noteToMidi Cs = 61
noteToMidi Db = 61
noteToMidi D = 62
noteToMidi Ds = 63
noteToMidi Eb = 63
noteToMidi E = 64
noteToMidi F = 65
noteToMidi Fs = 66
noteToMidi Gb = 66
noteToMidi G = 67
noteToMidi Gs = 68
noteToMidi Ab = 68
noteToMidi A = 69
noteToMidi As = 70
noteToMidi Bb = 70
noteToMidi B = 71


{- |
Absolute distance between two MIDI pitches.
-}
distance :: Pitch -> Pitch -> Int
distance n1 n2 = abs $ n1 - n2


{- |
Given a `Pitch` and a list of `Interval`s, returns a list of pitches each
distant an `Interval` from the starting `Pitch`.

It is basically a map over a list of `Interval`s.
-}
line :: Pitch -> [Interval] -> [Pitch]
line note = map (jump note)


{- |
Alters an interval (eg. @11 -> 11#@, @9 -> 9b@...); for example:

>>> alter Fourth Sharp
Tritone

Mostly used for chords alterations.
-}
alterInterval :: Interval -> Accident -> Interval
alterInterval int Flat  = toEnum (fromEnum int - 1)
alterInterval int Sharp = toEnum (fromEnum int + 1)



-- * Chords

-------------------------
-- Chords fundamentals --
-------------------------

{-
Data type for a chord.

It is important to underline that `Chord` is an __abstract__ representation of a chord;
that is to say, it is not made of MIDI pitches but of `Interval`s.

The act of translating this "abstract" chord to a "concrete" chord (ie. a collection of
MIDI pitches) is called /realization/ (see later).
-}
data Chord = Chord [Interval] deriving (Eq, Show)

{- |
How many notes a chord has.
-}
numNotes :: Chord -> Int
numNotes (Chord is) = length is

{- |
Returns the n-th inversion of a chord.
-}
inversion :: Int -> Chord -> Chord
inversion num (Chord is) = Chord $ rotate num is
  where rotate n xs = take (length xs) . drop n . cycle $ xs



-- ** Combinators for creating chords.

{- |
 Adds an interval on top.
-}
addOnTop :: Interval -> Chord -> Chord
addOnTop i (Chord is) = Chord $ is ++ [i]

{- |
 Alters an interval in the chord.
-}
alter :: Interval -> Accident -> Chord -> Chord
alter i accident (Chord is) = Chord $ mapEq (== i) (flip alterInterval accident) is 

{- |
 Removes an interval from the chord.
-}
remove :: Interval -> Chord -> Chord
remove interval (Chord is) = Chord $ filter (/= interval) is

{- |
 Sorts the chord. (useful?)
-}
sorted :: Chord -> Chord
sorted (Chord is) = Chord $ Data.List.sort is

{- |
 Combines chords combinators.
-}
(|>) :: (Chord -> Chord) -> (Chord -> Chord) -> Chord -> Chord
(|>) = flip (.)



-- ** Real Chords

majTriad = Chord [Unison, ThirdMaj, Fifth]
minTriad = Chord [Unison, ThirdMin, Fifth]

x_7    = addOnTop SeventhMin
x_maj7 = addOnTop SeventhMaj
x_9    = addOnTop NinthMaj
x_11   = addOnTop Eleventh
x_13   = addOnTop ThirteenthMaj
sus4   = remove ThirdMaj |> remove ThirdMin |> addOnTop Fourth |> sorted
soWhat = Chord [ThirdMaj, SixthMaj, NinthMaj, Twelfth, DoubleSeventhMaj]

{- |
 Realizes a chord, that is: given a Chord and a tonic, builds the Midi-pitch
representation of the chord.

>>> realizeChord majTriad 60
[60,64,67]

-}
realizeChord :: Chord -> Pitch -> [Pitch]
realizeChord (Chord is) tonic = line tonic is



-- * Sequences

-------------------------
-- Sequences of chords --
-------------------------

{- |
  A `Sequence` is a list of `Interval`s (the bass-line) and of `Chord`s.

  The bass-line has the form of a list of `Interval`; in a way, you could
describe a `Sequence` as a sequence of chords in a single tonality; or, also,
a sequence of chords that doesn't modulate.

  For every bass note at index i in `[Interval]`, the chord at index i of `[Chord]` will be built.

  An example of a ii-V-I sequence:

  >>> ii_v_i
  Sequence [Interval SecondMaj,Interval Fifth,Interval Unison]
           [Chord [Interval Unison,Interval ThirdMin,Interval Fifth,Interval SeventhMin],
            Chord [Interval Unison,Interval ThirdMaj,Interval Fifth,Interval SeventhMin],
            Chord [Interval Unison,Interval ThirdMaj,Interval Fifth,Interval SeventhMaj]]
-}
data Sequence = Sequence [Interval] [Chord] deriving (Show, Eq)

-- ** Sequences combinators

{- |
Remove the last index and chord of a sequence.
-}
removeLastSeq :: Sequence -> Sequence
removeLastSeq (Sequence is cs) = Sequence (init is) (init cs)

{- |
Substitute every occurrence of @(Interval, Chord)@ with another tuple.
-}
substitution :: (Interval, Chord) -> (Interval, Chord) -> Sequence -> Sequence
substitution old new (Sequence is cs) = Sequence (map fst subbed) (map snd subbed)
  where couples = zip is cs
        subbed  = mapEq (== old) (\_ -> new) couples

{- |
Tritone substitution.
-}
tritoneSub :: Sequence -> Sequence
tritoneSub = substitution
             (Fifth, x_7 majTriad)
             (SecondMin, x_7 majTriad)

{- |
So What? substitution; re-harmonizes maj7 chords on @I@ with soWhat chords
-}
soWhatSub :: Sequence -> Sequence
soWhatSub = substitution
            (Unison, x_maj7 majTriad)
            (Unison, soWhat)

(|>>|) :: (Sequence -> Sequence) -> (Sequence -> Sequence) -> Sequence -> Sequence
(|>>|) = (.)


{- |
  Realize a `Sequence`, that is: given a starting `Pitch`, build the chords on the intervals
relative to that midi pitch.

  For example:

  >>> realizeSeq 60 ii_v_i
  [[62,65,69,72],[67,71,74,77],[60,64,67,71]]

  Here, the @Interval Unison@ is set at pitch 60 (central C), so every interval is
calculated relative to 60 (eg. @Interval SecondMaj@ will be 62).
-}
realizeSeq :: Pitch -> Sequence -> [[Pitch]]
realizeSeq tonica (Sequence intervals chords) =
    zipWith (flip realizeChord) bassline chords
  where bassline = line tonica intervals

-- ** Shortest Sequence Realization

{- |
  Same as `realizeSeq`, but attempts to find the "nearest" substituions in the sequence.

  This means that, given two consecutive chords @C1@ and @C2@, @C2@ will be realized
with the inversion that minimizes the distance with @C1@.

  For example:

  >>> realizeSeqCompact 60 ii_v_i
  [[62,65,69,72],[62,65,67,71],[60,64,67,71]]

  Note: albeit being very easy to implement, this function still doesn't work properly.
-}
realizeSeqCompact :: Pitch -> Sequence -> [[Pitch]]
realizeSeqCompact tonica (Sequence intervals chords) = scanl1 shortestPathMidi realizedChords
  where bassline       = line tonica intervals
        realizedChords = zipWith (flip realizeChord) bassline chords

{- |
 All the inversions of a `Pitch`-chord
-}
inversionMidi :: [Pitch] -> [[Pitch]]
inversionMidi ns = concat $ map -- use non-deterministic monad? >>= on list
                   (\rot -> [rotateWith id rot ns,
                             rotateWith (+12) rot ns,
                             map (subtract 12) $ rotateWith (+12) rot ns])
                   [0..length ns]

{- |
 Find the "shortest" chord among the inversions of @C2@, relative to @C1@.
-}
shortestPathMidi :: [Pitch] -> [Pitch] -> [Pitch]
shortestPathMidi c1 c2 = inversions' !! indexMin
  where inversions' = inversionMidi c2
        ratedInversions = map (zipWith distance c1) inversions'
        indexMin  = fromJust $ elemIndex (minimum ratedInversions) ratedInversions


-- ** Some real sequences
v_i    = Sequence [Fifth, Unison]
                  [x_7 majTriad, x_maj7 majTriad]
ii_v_i = Sequence [SecondMaj, Fifth, Unison]
                  [x_7 minTriad, x_7 majTriad, x_maj7 majTriad]
ii_v_tritone = (removeLastSeq |>>| tritoneSub) ii_v_i



-- * Harmonies

{- |
  An `Harmony` is basically a collection of sequences, each of which has its own tonality.

  One big difference with `Sequence`s is that the @[Interval]@s are now relative to each other.
-}
data Harmony = Harmony [MelodicInterval] [Sequence] --tonalità - accordi
  deriving (Show, Eq)

{- |
A simple harmony with no modulation; it is basically a `Sequence`.
-}
emptyHarmony :: Sequence -> Harmony
emptyHarmony seq = Harmony [Up Unison] [seq]

circleFifths :: Harmony
circleFifths = Harmony ints seqs
  where ints = concat $ replicate 6 [Up Fourth, Down Fifth]
        seqs = replicate 12 ii_v_i

{- |
 Coltrane changes harmony.
-}
coltrane :: Harmony
coltrane = Harmony ints seqs
  where ints = concat $ replicate 3 [Up ThirdMaj, Down SixthMin]
        seqs = replicate 3 v_i

{- |
 Realize an `Harmony`.
-}
realizeHarmony :: Pitch -> Harmony -> [[Pitch]]
realizeHarmony note harmony = realizeHarmonyWith note realizeSeq harmony 

{- |
 Realize an `Harmony`, with the possibility to choose a custom sequence-realizing
function (eg. `realizeSeqCompact`).
-}
realizeHarmonyWith :: Pitch -> (Pitch -> Sequence -> [[Pitch]]) -> Harmony -> [[Pitch]]
realizeHarmonyWith note seqRealizer (Harmony is seq) =
  concatMap (\(tonica, s) -> seqRealizer tonica s) zipped
  where bassline = scanl (\acc i -> jump acc i) note is
        zipped   = zip bassline seq


------------------------------------------------------------------------------------

-- ** Midi Exporting

{- |
 Creates a complete MIDI file given a `MidiTrack`.
-}
createMidiFile :: MidiTrack -> MidiFile
createMidiFile track =
  MidiFile (MidiHeader MF0 1 (TPB 80))
           [track] --only 1 track!


data MidiOptions = MidiOptions
  { bpm :: Word32
  , shortestPath :: Maybe (Pitch -> Sequence -> [[Pitch]])
  , tuning :: Maybe HT.Temperament
}

{- |
 Given a list of chords (i.e. `[[Pitch]]`) returns a `MidiTrack`.
-}
makeTrack :: MidiOptions -> [[Pitch]] -> MidiTrack
makeTrack (MidiOptions bpm mayShortest mayTuning) chords =
  MidiTrack $ (trackPreamble bpm mayTuning) ++ concatMap midiChorder chords

  where seqRealizer = if isNothing mayShortest then realizeSeq else realizeSeqCompact


trackPreamble :: Word32 -> Maybe HT.Temperament -> [MidiMessage]
trackPreamble bpm tuning = 
  [ (0,MetaEvent (SetTempo $ bpmToMidiTempo bpm))
  , (0,MetaEvent (TimeSignature 4 2 24 8))        -- 4/4
  , (0,VoiceEvent RS_OFF (ProgramChange 0 0))     -- instrument
  ] ++ tuningMessage
  
  where tuningMessage = if isNothing tuning 
                        then [(0, MetaEvent (TextEvent SEQUENCE_NAME "Standard tuning!"))]
                        -- else HT.changeTuningMessage (fromJust tuning)
                        else HT.splitTuningMessage $ HT.changeTuningMessage (fromJust tuning)


{- |
 Translates [Pitch] into MIDI messages.
-}
midiChorder :: [Pitch] -> [MidiMessage]
midiChorder c = concat $ [notesOn, notesOff]
  where notesOn  = map (\pitch -> (0, VoiceEvent RS_OFF (NoteOn 0 (fromIntegral pitch) 100))) c
        notesOff = map (\pitch -> (64, VoiceEvent RS_OFF (NoteOff 0 (fromIntegral pitch) 0))) c



{- |
 BPM to MIDI tempo converter.
-}
bpmToMidiTempo :: Word32 -> Word32
bpmToMidiTempo bpm = 60_000_000 `div` bpm







writeMidiFile :: String -> MidiFile -> IO ()
writeMidiFile = writeMidi



exampleHarmony = circleFifths
examplePitches = realizeHarmonyWith (noteToMidi C) realizeSeq exampleHarmony
exampleTrack = makeTrack (MidiOptions 50 Nothing Nothing) examplePitches
exampleWrite = writeMidiFile "examples/chordsEq.mid" $ createMidiFile $ exampleTrack


      
--------------

